src/FastConv/FastConv.cpp

```diff
@@ -1,32 +1,192 @@
+//     fft->mergeRealImag(spectrumC, realA.data(), imagA.data());
+// }
+
+// Helper function: multiply two complex spectra, add result to output.
+static inline void addMultiplySpectra(float *output, const float *spectrumA, const float *spectrumB, int blockLength) {
+    // Spectrum layout: re(0),re(1),re(2),...,re(size/2),im(size/2-1),...,im(1)
+    output[0] += spectrumA[0] * spectrumB[0];
+    output[blockLength] += spectrumA[blockLength] * spectrumB[blockLength];
+    for (int i = 1; i < blockLength; i++) {
+        float realA = spectrumA[i];
+        float imagA = spectrumA[blockLength*2-i];
+        float realB = spectrumB[i];
+        float imagB = spectrumB[blockLength*2-i];
+        output[i] += realA * realB - imagA * imagB; // Real part
+        output[blockLength*2-i] += realA * imagB + imagA * realB; // Imaginary part
+    }
 }
 
-Error_t CFastConv::reset()
-{
+void FreqConvolution::process(float *output, const float *input, int length) {
+    int remainingInBlock;
+    // Fill and process as many complete blocks as we can.
+    while ((remainingInBlock = blockLength - indexInBlock) <= length) {
+        // Fill up the rest of the next input block.
+        std::copy_n(input, remainingInBlock, &inputBlocks[inputBlockIndex*blockLength*2 + indexInBlock]);
+        // Read out the rest of the last output block.
+        std::copy_n(&outputBlock[indexInBlock], remainingInBlock, output);
+
+        // We have another full input block for processing.
+        // Take the FFT in-place.
+        float *inputBlock = &inputBlocks[inputBlockIndex*blockLength*2];
+        std::fill(&inputBlock[blockLength], &inputBlock[blockLength*2], 0);
+        // NOTE: Unfortunately CFft adds some overhead here from copying buffers around,
+        // even though in this case we actually want the operation in-place.
+        // This is the best we can do without using LaszloFft directly.
+        fft->doFft(inputBlock, inputBlock);
+        // Save time-domain residue from last round & clear output block for accumulation.
+        std::copy_n(outputBlock.begin() + blockLength, blockLength, saved.begin());
+        std::fill(outputBlock.begin(), outputBlock.end(), 0);
+        // Sum results from circular convolutions (equivalent to linear convolutions due to zero-padding.)
+        // This is a kind of "higher-order" convolution of blocks.
+        for (int j = 0; j < numBlocks; j++) {
+            int idx = (inputBlockIndex - j + numBlocks) % numBlocks;
+            // NOTE: We don't need to re-scale the result here, because we already scaled `impulseResponseBlocks` in the constructor.
+            addMultiplySpectra(outputBlock.data(), &impulseResponseBlocks[j*blockLength*2], &inputBlocks[idx*blockLength*2], blockLength);
+        }
+        // Fourier Transform is linear, so we can just do the IFFT once after summing all the spectra.
+        fft->doInvFft(outputBlock.data(), outputBlock.data());
+        // Add back the residue from last time.
```

Comment:
The memory reuse of inputBlock and outputBlock is great!

Reply:
Thanks! While optimizing, I tried to reduce memory-copying as much as possible.